### PR TITLE
Fix rename/document-highlight/find reference in parameter property declaration

### LINKED ISSUE #4686: Unable to rename parameter properties
Rename `parent` to `parent_`.

```ts
class Foo {
  constructor(private parent) {
    parent
  }
}
```

expected

```ts
class Foo {
  constructor(private parent_) {
    parent_
  }
}
```

actual

```ts
class Foo {
  constructor(private parent_) {
    parent
  }
}
```

Visual Studio 2013

### PR DESCRIPTION
**Pull request by yuit**
fix #4686 rename/document highlighting/ find reference in parameter property declaration.

### CODE REVIEW

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -1444,10 +1444,7 @@ namespace ts {
 
             // If this is a property-parameter, then also declare the property symbol into the
             // containing class.
-            if (node.flags & NodeFlags.AccessibilityModifier &&
-                node.parent.kind === SyntaxKind.Constructor &&
-                isClassLike(node.parent.parent)) {
-
+            if (isPropertyParameterDeclaration(node)) {
```
Comment (DanielRosenwasser) [inline src/compiler/binder.ts]: `isParameterPropertyDeclaration`

File: src/compiler/types.ts [hunk 2]
```diff
@@ -1723,6 +1723,7 @@ namespace ts {
 
         getSymbolsInScope(location: Node, meaning: SymbolFlags): Symbol[];
         getSymbolAtLocation(node: Node): Symbol;
+        getSymbolOfParameterPropertyDeclaration(parameter: ParameterDeclaration, parameterName: string): Symbol[];
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: Symbol -> Symbols

File: src/compiler/utilities.ts [hunk 3]
```diff
@@ -2741,4 +2741,10 @@ namespace ts {
             }
         }
     }
+
+    export function isPropertyParameterDeclaration(node: ParameterDeclaration): boolean {
+        // If this is a property-parameter, then also declare the property symbol into the
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: I don't think this comment is relevant to the function

File: Jakefile.js [hunk 4]
```diff
@@ -924,7 +924,8 @@ function lintFileAsync(options, path, cb) {
 var lintTargets = compilerSources
     .concat(harnessCoreSources)
     .concat(serverCoreSources)
-    .concat(scriptSources);
+    .concat(scriptSources)
+    .concat([path.join(servicesDirectory,"services.ts")]);
```
Comment (yuit) [inline Jakefile.js]: It is currently only services file as to fix entire services folder for linting will be a large fix

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -428,6 +429,22 @@ namespace ts {
             // return undefined if we can't find a symbol.
         }
 
+        /**
+         * Get symbols that represent parameter-property-declaration as parameter and as property declaration
+         * @param parameter a parameterDeclaration node
+         * @param parameterName a name of the parameter to get the symbols for.
+         * @return a tuple of two symbols 
+         */
+        function getSymbolsOfParameterPropertyDeclaration(parameter: ParameterDeclaration, parameterName: string): [Symbol, Symbol] {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I wonder if the return type here should make it explicit over which is which (i.e.`{ parameterSymbol: Symbol; propertySymbol: Symbol; }`). Perhaps that's overkill.

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -428,6 +429,22 @@ namespace ts {
...
+         * @return a tuple of two symbols 
+         */
+        function getSymbolsOfParameterPropertyDeclaration(parameter: ParameterDeclaration, parameterName: string): [Symbol, Symbol] {
+            const constructoDeclaration = parameter.parent;
+            const classDeclaration = parameter.parent.parent;
+
+            const parameterSymbol = getSymbol(constructoDeclaration.locals, parameterName, SymbolFlags.Value);
+            const propertySymbol = getSymbol(classDeclaration.symbol.members, parameterName, SymbolFlags.Value);
+
+            return parameterSymbol && propertySymbol ? [parameterSymbol, propertySymbol] : undefined;
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Would it be appropriate to `Debug.fail` if you don't have both symbols?

**Comment by DanielRosenwasser** [comment]
I can't see any of the changes in `services.ts` due to your linting introduction. I think that change should have been done separately from this PR, so can we temporarily remove that change and cherry pick your lint fixes into a separate PR?

**Comment by yuit** [comment]
@DanielRosenwasser yes I can do that

File: src/compiler/checker.ts (hunk 6 above)
Comment (yuit) [inline src/compiler/checker.ts]: That probably will be appropriate

File: src/services/services.ts [hunk 7]
```diff
@@ -5967,6 +5967,15 @@ namespace ts {
                     }
                 }
 
+                // If the symbol.valueDeclaration is a property parameter declaration,
+                // we should include both parameter declaration symbol and property declaration symbol
+                // Parameter Declaration symbol is only visible within function scope, so the symbol is stored in contructor.locals.
+                // Property Declaration symbol is a member of the class, so the symbol is stored in its class Declaration.symbol.members
+                if (symbol.valueDeclaration && symbol.valueDeclaration.kind === SyntaxKind.Parameter &&
+                    isParameterPropertyDeclaration(<ParameterDeclaration>symbol.valueDeclaration)) {
+                    result = result.concat(typeChecker.getSymbolsOfParameterPropertyDeclaration(<ParameterDeclaration>symbol.valueDeclaration, symbol.name));
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: It think you can just write `result = typeChecker.getSymbolsOfParameter...`

Comment (DanielRosenwasser) [inline src/services/services.ts]: I wonder if it's permissible to repeat the same symbol, but I think it should be fine.

Comment (yuit) [inline src/services/services.ts]: How can it be the same symbol though?

**Comment by DanielRosenwasser** [comment]
:+1:

File: src/harness/fourslash.ts [hunk 8]
```diff
@@ -2768,6 +2768,10 @@ namespace FourSlashInterface {
             this.state.verifyCompletionListItemsCountIsGreaterThan(count, this.negative);
         }
 
+        public assertRangesEmpty(ranges: FourSlash.Range[]) {
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: This should be `assertHasRanges`

**Pull request merged**